### ISSUE TITLE
Option to deploy from build server

### ISSUE THREAD

**Issue by balduncle** [labzero/bootleg#239]
Would it be possible to add an option to deploy from the build server instead of first downloading to the invoker's machine and then pushing to the deployment targets?  My main issue is I have poor upload speeds and deploys pretty much kill my internet connectivity.

**Comment by brienw** [comment]
To add on to @StephaneRob's suggestion, you could use the `release_workspace` option, but instead of modifying the existing `build` task to use `scp` instead of `cp`, just add an `after: :build` task that does the additional step of `scp`ing the release from the `release_workspace` directly on `:build` over to the `release_workspace` directory on `:app` (assuming these are two different instances.)

**Comment by brienw** [comment]
Lets call this a workaround for now, I agree it would be nice to have a more formal support for this use-case, so I'm leaving this issue open as a feature request.

**Comment by balduncle** [comment]
Thanks for the advice.  I wasn't aware of the `release_workspace` option.  Per your comments and the readme it seems that having `release_workspace` set should cause the download step to be skipped.  This is not working for me.  I ensured that I'm running bootleg 0.7.0 and my build and app roles look quite similar to the example above, but every time I run `mix bootleg.update` it downloads the release to my laptop and then uploads from my laptop.

Any guesses as to what I might be doing wrong?

**Comment by balduncle** [comment]
Here's my redacted build role from `config/deploy.exs`.

```
role(
  :build,
  "BUILD_IP_ADDRESS,
  user: "BUILD_USER",
  identity: "PATH_TO_BUILD_KEY",
  workspace: "PATH_TO_BUILD_WORKSPACE",
  release_workspace: "PATH_TO_RELEASE_WORKSPACE"
)
```

Here's my redacted app role from `config/deploy/production.exs`.

```
role(
  :app,
  ["APP_IP_1", "APP_IP_2"],
  user: "APP_USER",
  identity: "PATH_TO_APP_KEY",
  workspace: "PATH_TO_WORKSPACE",
  release_workspace: "BUILD_USER@BUILD_IP_ADDRESS:PATH_TO_RELEASE_WORKSPACE"
)
```

**Comment by StephaneRob** [comment]
This feature was not released yet, so you have to change your deps to use the master version.

Moreover the `release_workspace` doesn't accept a remote path `BUILD_USER@BUILD_IP_ADDRESS:PATH_TO_RELEASE_WORKSPACE`, but you can create an extra task to scp the release from build remote to app remote(s) as @brienw explained above, and change the `release_workspace` of the app role to use a path in your app remote(s).

**Comment by balduncle** [comment]
Got it.  Thanks @StephaneRob.  My last struggle is figuring out how to scp from the build host to my other hosts.  Ideally I'd be able to forward my ssh keys from my bootleg-invoking host.  I'll keep researching this but if anyone has any advice there I'd appreciate it.  :)

**Comment by rjanja** [comment]
Another way would be to copy from the build server to the app servers. This has the same requirement as the solution above in that SSH access between the servers needs to be set up beforehand.

Here's an example of redeclaring the `copy_deploy_release` task, which is executed as part of the [remote deployment workflow](https://hexdocs.pm/bootleg/reference/workflow.html#deployment-workflow).

```elixir
role(:build, "build.example.com", release_workspace: "/opt/releases")
role(:app, ["app1.example.com", "app2.example.com"], release_workspace: "/opt/app")

task :copy_deploy_release do
  alias Bootleg.{UI, Config}

  build_role = Config.get_role(:build)

  build_release_path =
    Path.join(build_role.options[:release_workspace], "#{Config.version()}.tar.gz")

  app_role = Config.get_role(:app)
  app_release_path = app_role.options[:release_workspace]

  app_role
  |> Map.get(:hosts)
  |> Enum.each(fn bootleg_host ->
    host_name = bootleg_host.host.name
    UI.info("Copying release archive to #{host_name}")

    command =
      "scp -i ~/.ssh/remote-deploy-key #{build_release_path} #{host_name}:#{app_release_path}"

    IO.puts("-> #{command}")

    remote(:build, do: command)
  end)
end
```

Although I'm specifying the path on the build server for where the deploy key is, this should probably be left out and set in your `.ssh/config` on your build server - listing which options such as username, port and public key to use for each remote host.

```bash
$ mix bootleg.invoke copy_deploy_release
warning: task 'copy_deploy_release' is being redefined. The most recent definition will win, but this is probably not what you meant to do. The previous definition was at: ...[snip]
Copying release archive to app1.example.com
scp -i ~/.ssh/remote-deploy-key /opt/releases/0.1.0.tar.gz app1.example.com:/opt/app
Copying release archive to app2.example.com
scp -i ~/.ssh/remote-deploy-key /opt/releases/0.1.0.tar.gz app2.example.com:/opt/app
```

I think we could support this through documentation and examples, and maybe a way to redeclare tasks while suppressing warnings. Or we could add a deploy-from-build-server option and task similar to this example or the one above as a one-size-fits-most option.

**Comment by rjanja** [comment]
Closed by #308
